Investigate suspicious behavior of specialized task pools

mihails-strasuns-sociomantic opened the pull request:
Test case by @daniel-zullo-sociomantic : 

```D
import ocean.io.Stdout;
import ocean.task.TaskPool;

unittest
{
    static class SubTaskB : Task
    {
        int counter;

        override void run ()
        {
            Stdout.formatln("{}: Running B", this.counter);
            .wait(20);
            Stdout.formatln("{}: Finish B", this.counter);
        }

        void copyArguments (int counter)
        {
            this.counter = counter;
        }
    }

    static class SubTaskC : Task
    {
        int counter;

        override void run ()
        {
            Stdout.formatln("{}: Running C", this.counter);
            .wait(50);
            Stdout.formatln("{}: Finish C", this.counter);
        }

        void copyArguments (int counter)
        {
            this.counter = counter;
        }
    }

    static class SubTaskA : Task
    {
        static int counter;

        override void run ()
        {
            this.counter++;

            auto task_b = new SubTaskB();
            task_b.copyArguments(this.counter);
            theScheduler.await(task_b);
            theScheduler.processEvents();

            auto task_c = new SubTaskC();
            task_b.copyArguments(this.counter);
            theScheduler.await(task_c);
            theScheduler.processEvents();

            .wait(10);
        }

        void copyArguments ()
        {
        }
    }

    static class MainTask : Task
    {
        override void run ()
        {
            // suspend once because `await` safeguards against being run
            // before the scheduler starts
            theScheduler.processEvents();

            auto pool = new TaskPool!(SubTaskA);

            const NUM_RECORDS = 100;
            for (int i = 0; i < NUM_RECORDS; i++)
            {
                pool.start();
            }
        }
    }

    SchedulerConfiguration config;
    config.task_queue_limit = 10_000;
    config.worker_fiber_limit = 100;
    config.suspended_task_limit = 200;
    with (config)
    {
        specialized_pools = [
            PoolDescription(SubTaskB.classinfo.name, 102400),
            PoolDescription(SubTaskC.classinfo.name, 102400)
        ];
    }

    initScheduler(config);

    theScheduler.schedule(new MainTask);

    theScheduler.eventLoop();
}
```

daniel-zullo commented:
The `SubTaskC` is never run. Notice that same behaviour happens with or without specialized pools.
Also if the `SubTaskC` is not scheduled then the test runs smoothly

mihails-strasuns-sociomantic commented:
Problem manifests because specialized pools are not in fact used - https://github.com/sociomantic-tsunami/ocean/blob/acc8bb3a7de29911568ff374f3b4ae81e8c1436f/src/ocean/task/Scheduler.d#L411-L413

Because `await` calls explicitly `queue` and not `schedule`, none of registered specialized pools are considered.

I have to think a bit about best way to fix this.

daniel-zullo commented:
BTW there is a silly copy/pasta error calling the `copyArguments()` for `SubTaskC` in the test case. It should be:
```D
task_c.copyArguments(this.counter);
```

mihails-strasuns-sociomantic commented:
Fix attached. @daniel-zullo-sociomantic this fixes your example for me, please check if it helps with real app too.

nemanja-boric-sociomantic commented on src/ocean/task/Scheduler.d:
```diff
@@ -404,14 +404,21 @@ final class Scheduler : IScheduler
+            task.terminationHook({
+                finished_dg(task);
+            });
+        }
```
This is a matter for a separate commit.

nemanja-boric-sociomantic commented on src/ocean/task/internal/FiberPoolEager.d:
```diff
@@ -63,14 +63,24 @@ public class FiberPoolEager : FiberPool
-            this.recycle(fiber);
-            task.fiber = null;
-        });
+        task.terminationHook(&this.recycleHook);
```
Also a change for a separate commit?

nemanja-boric-sociomantic replied:
(together with the `recycleHook` definition)

mihails-strasuns-sociomantic commented:
Added release note

nemanja-boric-sociomantic commented:
The change itself LGTM.

mihails-strasuns-sociomantic committed:
Use class method instead of delegate

In FiberPoolEager delegate was using class `this` instead of stack
context `this` which may confuse DMD leading to wrong code gen. Making
it separate private class method ensure problem will not manifest.

mihails-strasuns-sociomantic committed:
Make await work with specialized task pools

Required a bit more convoluted logic with `schedule` instead of simply
calling `queue` as latter implies bypassing dedicated pool.

mihails-strasuns-sociomantic commented:
Tweaked commits

daniel-zullo commented:
>  please check if it helps with real app too.

I'll give it a shot asap (at the Dconf at the moment).

daniel-zullo commented:
I confirm a real app runs smoothly with the current patch. (I haven't reviewed the code for this patch though). I've also tested using template task classes as specialized pools and all went well too.
Thank you very much for taking care of this issue.

daniel-zullo reviewed (approved):
Code-wise LGTM

mihails-strasuns-sociomantic commented:
Ok, let's go ahead with this
